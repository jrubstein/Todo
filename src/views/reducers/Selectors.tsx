import { createSelector } from 'reselect'

const getItemsFilter = (state) =>  state.todos.items
const getEditingItemsFilter = (state) =>  state.todos.editingItem

export const getItemsSelector = () => {
    return createSelector(getItemsFilter, items => {
        return items
    })
}

export const getEditingItemsSelector = () => {
    return createSelector(getEditingItemsFilter, item => {
        return item
    })
}