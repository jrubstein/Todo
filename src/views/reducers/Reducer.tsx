import { initialState, TodoState } from './Store'
import { ADD_NEW_ITEM, EDIT_ITEM, SAVE_ITEM, DELETE_ITEM, TOGGLE_COMPLETE_ITEM } from '../actions/Actions'
import * as uuid from 'uuid/v4'

export const TodoReducer =  (state: TodoState = initialState.todos, action): TodoState => {
    switch (action.type) {
      case ADD_NEW_ITEM:
        const newItem = {
          id: uuid(),
          content: action.newValue,
          completed: false,
          createdAt: new Date()
        }
        return Object.assign({}, state, {
          items: [...state.items, ...[newItem]]
        })
      case EDIT_ITEM:
        return Object.assign({}, state, {
          editingItem: action.item
        })
      case DELETE_ITEM:
        return Object.assign({}, state, {
          items: state.items.filter(item => item.id !== action.id)
        })
      case SAVE_ITEM:
        const newItems = state.items.map(item => {
          if (action.id === item.id) {
            return {...item, ...{content: action.newContent}}
          }
          return item
        })

        return Object.assign({}, state, {
          items: newItems,
          editingItem: null
        })
      case TOGGLE_COMPLETE_ITEM:
        const toggledItems = state.items.map(item => {
          if (action.id === item.id) {
            return {...item, ...{completed: !item.completed}}
          }
          return item
        })
        return Object.assign({}, state, {
          items: toggledItems,
        })
      default:
        return state
    }
}