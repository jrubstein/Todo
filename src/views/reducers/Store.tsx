import { combineReducers, applyMiddleware, Store } from 'redux'
import { createStore } from 'redux'
import { TodoReducer } from './Reducer';

export interface Item  {
    id: string
    content: string
    completed: boolean
    createdAt: Date
}

export interface TodoState {
    items: Item[]
    editingItem?: Item
    showEditDialog: boolean
}

export interface State {
    todos: TodoState
}

export const initialState: State = {
    todos: {
        items: [],
        showEditDialog: false
    }
}

export const combinedReducers = combineReducers<State>({
    todos: TodoReducer
})

export const createNewStore = (init: State = initialState): Store<State> => {
    return createStore(combinedReducers , init)
    // For testing
    // !!(window || null) ? (window as any).__REDUX_DEVTOOLS_EXTENSION__ && (window as any).__REDUX_DEVTOOLS_EXTENSION__() : null)
}