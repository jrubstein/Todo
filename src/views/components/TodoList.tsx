import * as React from 'react'
import { State, Item } from '../reducers/Store';
import { connect } from 'react-redux';
import TodoItem from './TodoItem';
import { getItemsSelector } from '../reducers/Selectors'
import { sheet } from '../styles/base';

class TodoList extends React.Component<{items: Item[]}> {
    render() {
        return (
            <React.Fragment>
                <h2 className={sheet.classes.numberOfItems}>{this.props.items.length} Item(s)</h2>
                {this.props.items.map(item => <TodoItem key={item.id} item={item} /> )}
            </React.Fragment>
        )
    }
}

const mapStateToProps = () => {
    const getItems = getItemsSelector()
    return (state, props) => {
      return {
        items: getItems(state)
      }
    }
  }

export default connect(
    mapStateToProps,
    null,
)(TodoList)