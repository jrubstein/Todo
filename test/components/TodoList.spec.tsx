import { shallow, configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import { create } from 'react-test-renderer'
import TodoList from '../../src/views/components/TodoList'
import * as React from 'react'
import { createItem } from '../factories'
import { Provider } from 'react-redux';
import { createNewStore, initialState } from '../../src/views/reducers/Store';

configure({ adapter: new Adapter() })

describe('TodoList test', () => {
    it('Should create valid markup', () => {
        const component = create(
            <Provider store={createNewStore(initialState)}>
                <TodoList items={[createItem()]}/>
            </Provider>
        )
        expect(component).toMatchSnapshot()
    })
})
