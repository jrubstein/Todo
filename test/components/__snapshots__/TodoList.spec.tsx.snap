// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList test Should create valid markup 1`] = `
<h2
  className="numberOfItems-0-0-2"
>
  0
   Item(s)
</h2>
`;
